org.eclipse.scout.sdk.s2i/src/main/kotlin/org/eclipse/scout/sdk/s2i/util/SourceFolderHelper.kt

```diff
@@ -106,122 +98,210 @@ class SourceFolderHelper(val project: Project, val sourceFolder: SourceFolder, v
-            if (sharedSourceFolder == null) {
-                return null
+        private fun isTestSourceFolder(sourceFolder: SourceFolder?): Boolean = sourceFolderEndsWith(sourceFolder, ISourceFolders.TEST_JAVA_SOURCE_FOLDER) && sourceFolder?.rootType is JavaSourceRootType && sourceFolder.isTestSource
+
+        private fun isMainSourceFolder(sourceFolder: SourceFolder?): Boolean = sourceFolderEndsWith(sourceFolder, ISourceFolders.MAIN_JAVA_SOURCE_FOLDER) && sourceFolder?.rootType is JavaSourceRootType && !sourceFolder.isTestSource
+
+        private fun isGeneratedSourceFolder(sourceFolder: SourceFolder?): Boolean =
+            sourceFolderEndsWith(sourceFolder, IScoutSourceFolders.GENERATED_SOURCE_FOLDER) && sourceFolder?.rootType is JavaSourceRootType && !sourceFolder.isTestSource
+
+        private fun sourceFolderEndsWith(sourceFolder: SourceFolder?, suffix: String): Boolean {
+            if (sourceFolder == null) {
+                return false
             }
-            return findClosestDependentSourceFolder(project, ScoutTier.Client, false, sharedSourceFolder, scoutTierOfModule, findClasspathEntry)
+            return sourceFolder.file?.path?.endsWith(suffix) ?: false
         }
+    }
 
-        /**
-         * Finds the closest pair of [SourceFolder] and [IClasspathEntry] in the given [project] that is a [ScoutTier.Shared] for the given [clientOrServerSourceFolder].
-         */
-        fun findSharedSourceFolderPair(
-            project: Project,
-            clientOrServerSourceFolder: SourceFolder?,
-            scoutTierOfModule: ((Module) -> ScoutTier?)? = null,
-            findClasspathEntry: (VirtualFile?) -> IClasspathEntry?
-        ): Pair<SourceFolder, IClasspathEntry>? {
-            if (clientOrServerSourceFolder == null) {
-                return null
-            }
-            return findClosestSourceFolderDependency(project, ScoutTier.Shared, false, clientOrServerSourceFolder, scoutTierOfModule, findClasspathEntry)
+    fun tier(): ITier<*>? = m_tier.computeIfAbsentAndGet {
+        tierOfModule(sourceFolder.contentEntry.rootModel.module)
+    }
+
+    fun classpathEntry(): IClasspathEntry? = m_classpathEntry.computeIfAbsentAndGet {
+        findClasspathEntry(sourceFolder.file)
+    }
+
+    fun sourceFolder(tier: ITier<*>): IClasspathEntry? = sourceFolderBundle(tier)?.sourceFolderPair()?.second
+
+    fun testSourceFolder(tier: ITier<*>): IClasspathEntry? = sourceFolderBundle(tier)?.testSourceFolderPair()?.second
+
+    fun mainTestSourceFolder(tier: ITier<*>): IClasspathEntry? = sourceFolderBundle(tier)?.mainTestSourceFolder()
+
+    fun generatedSourceFolder(tier: ITier<*>): IClasspathEntry? = sourceFolderBundle(tier)?.generatedSourceFolder()
+
+    private fun sourceFolderBundle(tier: ITier<*>): SourceFolderBundle? {
+        // add start sourceFolderBundle
+        m_tierSourceFolderBundleMap.computeIfAbsent(tier()) {
```

Comment:
what happens if tier() returns null (if no Scout is on the classpath)? Maybe add testcase for it?